src/engine/guidance/post_processing.cpp

```diff
@@ -20,194 +22,248 @@ namespace guidance
-    {
-        BOOST_ASSERT(canMergeTrivially(destination, source));
-        return detail::forwardInto(destination, source);
-    }
-    if (source.turn_instruction.type == TurnType::Suppressed)
-    {
-        return detail::forwardInto(destination, source);
-    }
-    if (source.turn_instruction.type == TurnType::StayOnRoundabout)
-    {
-        return detail::forwardInto(destination, source);
-    }
-    if (entersRoundabout(source.turn_instruction))
-    {
-        return detail::forwardInto(destination, source);
-    }
+    destination.duration += source.duration;
+    destination.distance += source.distance;
+    destination.geometry_begin = std::min(destination.geometry_begin, source.geometry_begin);
+    destination.geometry_end = std::max(destination.geometry_end, source.geometry_end);
     return destination;
 }
 
 } // namespace detail
 
-void print(const std::vector<std::vector<PathData>> &leg_data)
+void print(const std::vector<RouteStep> &steps)
 {
     std::cout << "Path\n";
-    int legnr = 0;
-    for (const auto &leg : leg_data)
+    int segment = 0;
+    for (const auto &step : steps)
     {
-        std::cout << "\tLeg: " << ++legnr << "\n";
-        int segment = 0;
-        for (const auto &data : leg)
-        {
-            const auto type = static_cast<int>(data.turn_instruction.type);
-            const auto modifier = static_cast<int>(data.turn_instruction.direction_modifier);
+        const auto type = static_cast<int>(step.maneuver.instruction.type);
+        const auto modifier = static_cast<int>(step.maneuver.instruction.direction_modifier);
 
-            std::cout << "\t\t[" << ++segment << "]: " << type << " " << modifier
-                      << " exit: " << data.exit << "\n";
-        }
+        std::cout << "\t[" << ++segment << "]: " << type << " " << modifier
+                  << " Duration: " << step.duration << " Distance: " << step.distance
+                  << " Geometry: " << step.geometry_begin << " " << step.geometry_end
+                  << " exit: " << step.maneuver.exit
+                  << " Intersections: " << step.maneuver.intersections.size() << " [";
+
+        for (auto intersection : step.maneuver.intersections)
+            std::cout << "(" << intersection.duration << " " << intersection.distance << ")";
+
+        std::cout << "] name[" << step.name_id << "]: " << step.name << std::endl;
     }
-    std::cout << std::endl;
 }
 
-std::vector<std::vector<PathData>> postProcess(std::vector<std::vector<PathData>> leg_data)
+// Every Step Maneuver consists of the information until the turn.
+// This list contains a set of instructions, called silent, which should
+// not be part of the final output.
+// They are required for maintenance purposes. We can calculate the number
+// of exits to pass in a roundabout and the number of intersections
+// that we come across.
+
+std::vector<RouteStep> postProcess(std::vector<RouteStep> steps)
```

Comment:
This is a 200 loc function and is somewhat hard to follow --- I would love if we could refactor it a bit